packages/flutter/lib/src/widgets/scrollbar.dart

```diff
@@ -0,0 +1,208 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'package:flutter/animation.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+import 'scroll_metrics.dart';
+
+const double _kMinThumbExtent = 18.0;
+
+/// A [CustomPainter] for painting scrollbars.
+///
+/// Unlike [CustomPainter]s that sub-classes [CustomPainter] and only repaints
+/// when [shouldRepaint] returns true (which requires this [CustomPainter] to
+/// be rebuilt), this painter has the added optimization of repainting and not
+/// rebuilding when:
+///
+///  * the scroll position changes; and
+///  * when the scrollbar fades away.
+///
+/// See also:
+///
+///  * [Scrollbar] for a widget showing a scrollbar around a [Scrollable] in the
+///    Material Design style.
+///  * [CupertinoScrollbar] for a widget showing a scrollbar around a
+///    [Scrollable] in the iOS style.
+class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
+  /// Creates a scrollbar with customizations given by construction arguments.
+  ScrollbarPainter({
+    @required this.color,
+    @required this.textDirection,
+    @required this.thickness,
+    @required this.fadeoutOpacityAnimation,
+    this.mainAxisMargin: 0.0,
+    this.crossAxisMargin: 0.0,
+    this.radius,
+    this.minLength: _kMinThumbExtent,
+  }) : assert(color != null),
+       assert(textDirection != null),
+       assert(thickness != null),
+       assert(fadeoutOpacityAnimation != null),
+       assert(mainAxisMargin != null),
+       assert(crossAxisMargin != null),
+       assert(minLength != null) {
+    fadeoutOpacityAnimation.addListener(notifyListeners);
```

Comment:
when do you removeListener?

Reply:
I think you need to override dispose() and call removeListener there.

Also, the docs need to talk about how it's important to call dispose() on this object when you're done with it.

Reply:
Added doc about needing to dispose.
ChangeNotifier.dispose already more or less does removeListener.

Reply:
How does ChangeNotifier know about fadeoutOpacityAnimation?

Reply:
ah you're right. Sorry, I mixed the 2 layers of listeners up.